### PR TITLE
mon/osdmonitor: fix incorrect output of "osd df" due to osd out

### PR DESCRIPTION
**Pull request by xiexingguo**
If an osd is automatically marked as out, the output of "osd df"
    is not right, as follow:
    
    -5 10.00999        -  5586G  2989G  2596G     0    0     host ceph192-9-9-8
    11  0.90999  1.00000   931G   542G   388G 58.25 0.99         osd.11
    14  0.90999  1.00000   931G   530G   400G 56.97 0.97         osd.14
    20  0.90999  1.00000   931G   716G   214G 76.99 1.31         osd.20
    22  0.90999  1.00000   931G   477G   453G 51.29 0.87         osd.22
    26  0.90999        0      0      0      0     0    0         osd.26
    28  0.90999  1.00000   931G   587G   343G 63.09 1.07         osd.28
    30  0.90999  1.00000   931G   602G   328G 64.75 1.10         osd.30
    16  0.90999  1.00000   931G   589G   341G 63.34 1.08         osd.16
    18  0.90999  1.00000   931G   530G   400G 56.93 0.97         osd.18
    24  0.90999  1.00000   931G   202G   728G 21.77 0.37         osd.24
    32  0.90999  1.00000   931G   477G   454G 51.23 0.87         osd.32

Two problems are identified from the above output:

1. the total capacity(total, total used, total avial) 
only includes osd.32, osd.24, osd.18, osd.16, osd.30, osd.28, and other
healthy osds such as osd.11, osd.14 etc. are excluded.

2. the average utilization/deviation are forced resetted.

Fixes: http://tracker.ceph.com/issues/16706

### CODE REVIEW

**Comment by xiexingguo** [comment]
Local verified, as below:
```
-9 3.42000        -  2937G   135M  2937G 0.00  0.58     host ceph192-9-9-236     
 3 0.53999  1.00000   548G 35064k   548G 0.01  0.78         osd.3                
 5 0.53999  1.00000   548G 34560k   548G 0.01  0.77         osd.5                
 6 0.53999        0      0      0      0    0     0         osd.6                
 7 0.89999  1.00000   920G 34384k   920G 0.00  0.46         osd.7                
 4 0.89999  1.00000   920G 34408k   920G 0.00  0.46         osd.4                
              TOTAL  2996G   239M  2996G 0.01

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -796,14 +796,20 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
 
   bool get_osd_utilization(int id, int64_t* kb, int64_t* kb_used,
 			   int64_t* kb_avail) const {
+    // we always return true for OSDs as the specific OSDs can be out,
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: could you be more specific in the commit message, "chaos" is sort of vague, imho.

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -796,14 +796,20 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
...
+      *kb = 0;
+      *kb_used = 0;
+      *kb_avail = 0;
+      return true;
+    }
     *kb = p->second.kb;
     *kb_used = p->second.kb_used;
     *kb_avail = p->second.kb_avail;
-    return *kb > 0;
+    return true;
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: why this change?

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -823,7 +823,7 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
 
     for (int k = osdmap->crush->get_bucket_size(id) - 1; k >= 0; k--) {
       int item = osdmap->crush->get_bucket_item(id, k);
-      int64_t kb_i = 0, kb_used_i = 0, kb_avail_i;
+      int64_t kb_i = 0, kb_used_i = 0, kb_avail_i = 0;
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: good catch!

**Comment by xiexingguo** [comment]
I'll give an elaborate explanation when I back to work:)

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: updated.

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: I thought the out OSD should be caught by the condition https://github.com/ceph/ceph/pull/10308/files#diff-0a5db46a44ae9900e226289a810f10e8R803 at first, but it turned out the out OSD's statistics is zeroed. So this change is necessary.

**Comment by xiexingguo** [comment]
The change is locally tested and verified, as below:

```
-9 3.42000        -  2937G   135M  2937G 0.00  0.58     host ceph192-9-9-236     
 3 0.53999  1.00000   548G 35064k   548G 0.01  0.78         osd.3                
 5 0.53999  1.00000   548G 34560k   548G 0.01  0.77         osd.5                
 6 0.53999        0      0      0      0    0     0         osd.6                
 7 0.89999  1.00000   920G 34384k   920G 0.00  0.46         osd.7                
 4 0.89999  1.00000   920G 34408k   920G 0.00  0.46         osd.4                
              TOTAL  2996G   239M  2996G 0.01  
```


The bucket statistics can reflect the summarized statistics of all OSDs under correctly even if one or more OSDs are offline and marked as out.

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: if an OSD could be marked out, can we just check its state in `get_bucket_utilization()` explicitly using `osdmap.is_out(id)`?

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: This is my first thought.

But there is no reason we stop to process the whole bucket if any one of the OSDs under this bucket is abnormal because they are out or broken or something alike. And if they are, reset their stats will be fair enough.

Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: > But there is no reason we stop to process the whole bucket if any one of the OSDs under this bucket is abnormal

agreed. and could you be specific why checking `osdmap.is_out(id)` in `get_bucket_utilization()`  does not work, or is a bad idea?

we are tackling two cases here:

1. we should not stop calc a bucket's usage when running into an OSD gets marked out: in this case, i'd suggest check `osdmap.is_out(id)` explicitly. actually marking an OSD out will reset its stat to 0, see `PGMonitor::check_osd_map()` and `PGMap::apply_incremental()`
1. in other cases, i just want to be sure it's expected behavior, if it's not, i don't want to cover it up using this fix. if we destroy an OSD, its stat will be removed from `PGMap::osd_stat`, see `PGMonitor::check_osd_map()`, in that case, why is its parent bucket still holding it?
```c++
      if (p->second & CEPH_OSD_EXISTS) {
	// whether it was created *or* destroyed, we can safely drop
	// it's osd_stat_t record.
	dout(10) << __func__ << "  osd." << p->first
	         << " created or destroyed" << dendl;
	pending_inc.rm_stat(p->first);

	// and adjust full, nearfull set
	pg_map.nearfull_osds.erase(p->first);
	pg_map.full_osds.erase(p->first);
      }
```

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: > could you be specific why checking osdmap.is_out(id) in get_bucket_utilization() does not work, or is a bad idea?

@tchaikov  That's not true. I think the `is_out()` method shall work too(It is my first at all solution actually:-). But I have the following concerns too:

- I think https://github.com/ceph/ceph/pull/10308/files?diff=unified#diff-0a5db46a44ae9900e226289a810f10e8R803 is unreachable, as we have done sanity check at:

```
  void dump_stray(F *f) {
    for (int i = 0; i < osdmap->get_max_osd(); i++) {
      if (osdmap->exists(i) && !this->is_touched(i))
	    dump_item(CrushTreeDumper::Item(i, 0, 0), f);
    }
  }
```

- I am afraid there is other possibilities that an OSD report zeroed stats, please check:

```
void OSDService::update_osd_stat(vector<int>& hb_peers)
{
  Mutex::Locker lock(stat_lock);

  osd_stat.hb_in.swap(hb_peers);
  osd_stat.hb_out.clear();

  osd->op_tracker.get_age_ms_histogram(&osd_stat.op_queue_age_hist);

  // fill in osd stats too
  struct store_statfs_t stbuf;
  int r = osd->store->statfs(&stbuf);
  if (r < 0) {
    derr << "statfs() failed: " << cpp_strerror(r) << dendl;

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: > I am afraid there is other possibilities that an OSD report zeroed stats, please check:
> 
> void OSDService::update_osd_stat(vector<int>& hb_peers)
> {
>   Mutex::Locker lock(stat_lock);
> 
>   osd_stat.hb_in.swap(hb_peers);
>   osd_stat.hb_out.clear();
> 
>   osd->op_tracker.get_age_ms_histogram(&osd_stat.op_queue_age_hist);
> 
>   // fill in osd stats too
>   struct store_statfs_t stbuf;
>   int r = osd->store->statfs(&stbuf);
>   if (r < 0) {
>     derr << "statfs() failed: " << cpp_strerror(r) << dendl;

Sorry, not zeroed but stale OSD stats...
So I'll try in your way.

**Comment by xiexingguo** [comment]
Upgraded. @tchaikov  Can you look again? Thanks!

File: src/mon/OSDMonitor.cc [hunk 4]
```diff
@@ -796,6 +796,15 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
 
   bool get_osd_utilization(int id, int64_t* kb, int64_t* kb_used,
 			   int64_t* kb_avail) const {
+    if (osdmap->is_out(id)) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: can we do this in `get_bucket_utilization()`?

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: sure, done.

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -808,8 +808,18 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
 
   bool get_bucket_utilization(int id, int64_t* kb, int64_t* kb_used,
 			      int64_t* kb_avail) const {
-    if (id >= 0)
+    if (id >= 0) {
+      if (osdmap->is_out(id)) {
+        // return true if current OSD is marked as out,
+        // so we can continue to process other OSDs under this bucket.
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: i don't think this comment is necessary, as the code is self-explained.

File: src/mon/OSDMonitor.cc [hunk 6]
```diff
@@ -808,8 +808,18 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
...
+        *kb = 0;
+        *kb_used = 0;
+        *kb_avail = 0;
+        return true;
+      }
+
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: could you remove this empty line?

File: src/mon/OSDMonitor.cc (hunk 3 above)
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: i just realized this is not a bug. as we already initialize them. could you remove the "correctly" in the commit message if you are with me.

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: okay.

**Comment by tchaikov** [comment]
and @xiexingguo , could you please file a ticket on tracker and connect ab685f7a2c5edf136a621b3c593404b3d7f0b616 with it, so we can backport it to jewel ?

**Comment by xiexingguo** [comment]
All addressed. Thanks @tchaikov

**Comment by tchaikov** [comment]
lgtm

**Comment by xiexingguo** [comment]
retest this please

**Comment by tchaikov** [comment]
tested at

- http://pulpito.ceph.com/kchai-2016-07-24_02:17:13-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2016-07-24_20:41:33-rados-wip-kefu-testing---basic-mira/

**Comment by xiexingguo** [comment]
@tchaikov Thanks for the testing and merging, kefu.
